src/fdc/MSXtar.cc

```diff
@@ -107,10 +117,18 @@ void MSXtar::parseBootSector(const MSXBootSector& boot)
 	chrootSector = rootDirStart;
 	rootDirLast = rootDirStart + nbRootDirSectors - 1;
 
+	// Whether to use FAT16 is strangely implicit, it must be derived from the
+	// cluster count being higher than 0xFF4. Let's round that up to 0x1000,
+	// since FAT12 uses 1.5 bytes per cluster its maximum size is 12 sectors,
+	// whereas FAT16 uses 2 bytes per cluster so its minimum size is 16 sectors.
+	// Therefore we can pick the correct FAT type with sectorsPerFat >= 14.
+	unsigned fat16Threshold = (0x1000 * 3 / 2 + 0x1000 * 2) / 2 / SECTOR_SIZE;
+	fat16 = sectorsPerFat >= fat16Threshold;
+
 	// Bypass max cluster check in sectorToCluster
 	maxCluster = std::numeric_limits<decltype(maxCluster)>::max();
-	maxCluster = sectorToCluster(boot.nrSectors);
-	maxCluster = std::min(maxCluster, BAD_FAT);
+	maxCluster = sectorToCluster(nrSectors);
+	maxCluster = std::min(maxCluster, fat16 ? MAX_FAT16 : MAX_FAT12);
```

Comment:
This changes from `BAD_FAT` to `MAX_FAT`. That was probably a bug in the old code which you fixed?

Maybe we should add some comments to clarify:
* Is `MAX_FAT` inclusive or exclusive. IOW is it the last still valid or the first invalid number.
* Similar for the `maxCluster` variable, is it inclusive or exclusive?
Hmm, but from the formula `sectorToCluster(nrSectors)` I get the impression that `maxCluster` is exclusive because also `nrSectors` is exclusive.

Reply:
Yes indeed.

I added `BAD_FAT` in an earlier submit (before then it was unclamped), but since then I got more detail about what the maximum legal (non-special) cluster number is. `MAX_FAT12` is the maximum cluster number + 1 (so, exclusive) for FAT12 and ditto for `MAX_FAT16`. And, because of the encoding of cluster numbers, the maximum number of clusters is `MAX_FAT*` - 2.

The `maxCluster` variable is also exclusive, that is, it is the maximum cluster number + 1 or the maximum cluster count + 2.

It was also not clear to me at first, so I can see where you’re coming from, I’ll add some comments. Or maybe I can do something with the name or the way they’re used, e.g. renaming to `clusterCount` and then doing +2 when it’s used to compare with a cluster number. Counts are unambiguously exclusive, after all.